File: rzcommon.c
#include "rzcommon.h"

static void rz_data_from_report(const struct rz_report *report, uint8_t *data) {
    uint8_t *data_start = data;

    *data = 0x00; //Report ID
    data++;

    const uint8_t NUM_PARAMS = report->params_len + 1; //params_len + 1 for sub_cmd
    const uint8_t DATA_HDR[7] = {0x00, report->id, 0x00,  0x00, 0x00, NUM_PARAMS, 0x03}; //Last 0x03 reserved byte

    const uint8_t DATA_CMD[2] = {report->cmd, report->sub_cmd};

    memcpy(data, DATA_HDR, 7);
    data += 7;

    memcpy(data, DATA_CMD, 2);
    data += 2;

    memcpy(data, report->params, report->params_len);
    data += report->params_len;

    while (data != data_start + ((RZ_REPORT_LEN + 1) * sizeof(uint8_t))) {
        *data = 0;
        data++;
    }

    data_start[RZ_REPORT_LEN + 1 - 2] = rz_calculate_crc(data_start);
}

uint8_t rz_calculate_crc(const uint8_t *data) {
    uint8_t crc = 0;
    data++; //Skip report ID

    for (int i = 2; i < 88; i++) {
        crc ^= data[i];
    }

    return crc;
}

int rz_send_report(hid_device *dev, const struct rz_report *report) {
    uint8_t data[RZ_REPORT_LEN + 1];

    rz_data_from_report(report, data);

    return hid_send_feature_report(dev, data, RZ_REPORT_LEN + 1);
}

bool rz_set_brightness(hid_device *dev, const float brightness) {
    uint8_t PARAMS[2] = {0x05, (brightness * 255)};

    struct rz_report report;
    report.id = 0x1f;
    report.cmd = 0x03;
    report.sub_cmd = 0x01;
    report.params = PARAMS;
    report.params_len = 2;

    return rz_send_transfer(dev, &report) != -1;
}

File: rzcommon.rs
use std::ptr::null;

use rusb::{DeviceHandle, GlobalContext};

use crate::rzdevices::*;

enum RzDeviceType {
    MOUSE,
    KEYBOARD,
    KRAKEN,
    ACCESSORY
}

pub struct RzDevice {
    pub usb_dev: Option<DeviceHandle<GlobalContext>>,
    pub pid: u16,
    pub w_index: u16,
    pub dev_type: RzDeviceType
}

pub struct RzReport {
    pub id: u8,
    pub cmd: u8,
    pub sub_cmd: u8,
    pub params: Vec<u8>
}

const RZ_VENDOR_ID: u16 = 0x1532;
const RZ_REPORT_LEN: usize = 90;

fn rz_calculate_crc(data: &[u8]) -> u8 {
    let mut crc: u8 = 0;

    for i in 2..88 {
        crc ^= data[i];
    }

    return crc;
}

impl From<&RzReport> for [u8; RZ_REPORT_LEN] {
    fn from(report: &RzReport) -> Self {
        let mut data: [u8; RZ_REPORT_LEN] = [0; RZ_REPORT_LEN];

        let num_params: u8 = (report.params.len() as u8) + 1;
        let data_hdr: [u8; 7] = [0x00, report.id, 0x00, 0x00, 0x00, num_params, 0x03];

        let data_cmd: [u8; 2] = [report.cmd, report.sub_cmd];

        data[0..7].copy_from_slice(&data_hdr);
        data[7..9].copy_from_slice(&data_cmd);

        for (i, param) in report.params.iter().enumerate() {
            data[9 + i] = *param;
        }

        data[RZ_REPORT_LEN - 2] = rz_calculate_crc(&data);

        return data;
    }
}

impl Default for RzDevice {
    fn default() -> Self {
        RzDevice {
            pid: Default::default(),
            usb_dev: None,
            w_index: Default::default(),
            dev_type: RzDeviceType::KEYBOARD
        }
    }
}

impl RzDevice {
    fn get_w_index(&self) -> u16 {
        match(self.pid) {
            RZ_PID_BLACKWIDOW_CHROMA_V2 => 0x02,
            _ => 0x00
        }
    }

    pub fn open(&mut self, pid: u16) {
        self.pid = pid;
        self.usb_dev = rusb::open_device_with_vid_pid(RZ_VENDOR_ID, pid);
        self.w_index = self.get_w_index();
        self.dev_type = RzDeviceType::KEYBOARD; //TODO: Implement this

        self.usb_dev.claim_interface(self.w_index as u8);
    }

    pub fn close(&self) {
        self.usb_dev.release_interface(self.w_index as u8);
        self.close();
    }

    pub fn send_report(&self, report: &RzReport) -> bool {
        let mut data: [u8; RZ_REPORT_LEN] = report.into();

        return self.usb_dev.write_control(
            0x21, 
            0x09, 
            0x300, 
            self.w_index, 
            &data,
            std::time::Duration::from_millis(2000)
        ).unwrap() == RZ_REPORT_LEN;
    }
}